### ISSUE TITLE
OptimzeAsync didn't remove files

### ISSUE THREAD

**Issue by edwin-jsi** [thebracket/LanceDbCSharp#16]
var optimizeResult = await table1.OptimizeAsync(TimeSpan.Zero);
    System.Console.WriteLine("Optimize Result: " + optimizeResult.Compaction + optimizeResult.Prune);

the optimizeResult shows nothing has been changed.

with python api, if I run 
    await tbl.optimize(cleanup_older_than=timedelta(0))
I can see the files in the data folder will be merged. i.e. some files will be deleted.

**Comment by edwin-jsi** [comment]
used following code to create multiple files before calling OptimizeAsync

    string[] columns = new[] { "id" };    
    ILanceMergeInsertBuilder builder = await table1.MergeInsertAsync(columns);
    numEntries = 100000;
    for (int i = 0; i < 10; i++)
    {
        IEnumerable<RecordBatch> records = GetBatches(numEntries, i * numEntries);
        await builder.WhenMatchedUpdateAll().WhenNotMatchedInsertAll().ExecuteAsync(records);
        System.Console.WriteLine($"Table 1 row count (expected {numEntries * (i+1)}) actual {table1.CountRows()}" );
    }

**Comment by edwin-jsi** [comment]
the following code will have same result.
var optimizeResult = await table1.OptimizeAsync(TimeSpan.FromDays(0));

**Comment by thebracket** [comment]
Good catch! I'd missed actually passing those parameters onto LanceDb. Corrected, and now there's a sync and async unit test to verify that it does something.

**Comment by thebracket** [comment]
So you *also* want to run a compact when that fires? (Pruning and compacting are separate calls). Ok.

**Comment by thebracket** [comment]
Ok, with the most recent commit that should be what you want.

**Comment by thebracket** [comment]
Ok, the `TestOptimizeRowsAsync` now uses multiple batches, to create multiple transactions - and then calls optimize. I found a behavior that was missing in the Rust side - calling `checkout_latest`, which seems to have taken care of the issue (Lance appears to hold onto checked out editions until they are done - which makes sense).

Adding 10 batches of 8 each:

*Pre compaction*:

```
find /tmp/test_table_optimize_async
.
./table1.lance
./table1.lance/_transactions
./table1.lance/_transactions/0-852c7ac8-87ec-4115-be78-847e5b1455fe.txn
./table1.lance/_transactions/1-cf001d8c-f7a1-4d04-ac76-603dfeef0c70.txn
./table1.lance/_transactions/4-8e904cf9-524e-4cf8-9461-203e68289b32.txn
./table1.lance/_transactions/7-1f98c798-d591-4083-b6e3-d5df627f9a31.txn
./table1.lance/_transactions/8-81fadb1b-f26a-4a41-beb8-e63932efa440.txn
./table1.lance/_transactions/2-d3cddaf9-e7ef-461d-8e22-006278e30b1a.txn
./table1.lance/_transactions/3-92954ed1-b7ca-4905-956c-9984a018fb95.txn
./table1.lance/_transactions/5-4390b71e-8a7c-46ea-9789-21a72bcf703d.txn
./table1.lance/_transactions/9-928ba144-5594-4660-ba2b-3d095c4648ff.txn
./table1.lance/_transactions/10-234441ce-72e2-4583-b664-5ce6bb625047.txn
./table1.lance/_transactions/6-8a3c0114-9d8e-4f37-b756-d7e8ff4bc371.txn
./table1.lance/data
./table1.lance/data/b04ae949-4724-45ca-8ec2-433f71577046.lance
./table1.lance/data/81c973d6-4d23-44a5-b438-1f2956d01706.lance
./table1.lance/data/3660ad94-e861-46bc-9960-25e57e02f1b0.lance
./table1.lance/data/476cdd1e-2d42-40b9-992c-93ac131ccf93.lance
./table1.lance/data/d626e97f-d221-4de0-9c45-3c2ca06571c5.lance
./table1.lance/data/b6e68664-38bf-4091-b1cb-58247ca31d77.lance
./table1.lance/data/461bcb86-0c02-4291-84db-19336ac61952.lance
./table1.lance/data/9c0dcd8e-88c3-4971-a4bc-3357b35911e4.lance
./table1.lance/data/0fe57e50-79e5-4871-a5cf-a3119cb04df8.lance
./table1.lance/data/b8eb0471-d734-40da-bf61-58e62086490e.lance
./table1.lance/_versions
./table1.lance/_versions/9.manifest
./table1.lance/_versions/2.manifest
./table1.lance/_versions/11.manifest
./table1.lance/_versions/5.manifest
./table1.lance/_versions/6.manifest
./table1.lance/_versions/8.manifest
./table1.lance/_versions/7.manifest
./table1.lance/_versions/3.manifest
./table1.lance/_versions/1.manifest
./table1.lance/_versions/4.manifest
./table1.lance/_versions/10.manifest
```

*Post compaction*:

```
find /tmp/test_table_optimize_async
.
./table1.lance
./table1.lance/_transactions
./table1.lance/_transactions/10-234441ce-72e2-4583-b664-5ce6bb625047.txn
./table1.lance/data
./table1.lance/data/b6e68664-38bf-4091-b1cb-58247ca31d77.lance
./table1.lance/_versions
./table1.lance/_versions/11.manifest
```

So it's definitely removed plenty of files.

**Comment by edwin-jsi** [comment]
https://lancedb.github.io/lancedb/python/python/#lancedb.table.Table.optimize

Optimization covers three operations:

Compaction: Merges small files into larger ones
Prune: Removes old versions of the dataset
Index: Optimizes the indices, adding new data to existing indices

do you optimize indices while calling optimize?

**Comment by thebracket** [comment]
Ok, that last patch adds an explicit index optimization call to the list when calling optimize with parameters (the no params case uses the "All" command which performs all steps, but doesn't include pruning parameters)

**Comment by edwin-jsi** [comment]
when I tested it, I still have more files left.
before optimize,
# ls -l ./data
total 20528
-rw-r--r-- 1 root root 2100796 Jan  6 20:33 0c602541-d597-4309-9960-f864b9c18d37.lance
-rw-r--r-- 1 root root 2100796 Jan  6 20:33 1004682c-c808-4f22-a9e6-bfb74ca0214c.lance
-rw-r--r-- 1 root root 2100796 Jan  6 20:33 587e1021-9d31-4865-aa36-358ba64e777a.lance
-rw-r--r-- 1 root root 2100796 Jan  6 20:33 628aa704-5647-4138-87ce-4d777a20e20d.lance
-rw-r--r-- 1 root root     848 Jan  6 20:33 6ca6ef1d-611b-48c1-bfcb-5a2a3eab7755.lance
-rw-r--r-- 1 root root 2100796 Jan  6 20:33 79bc4fcf-f7c2-4910-b3e9-2a3a78ab34f1.lance
-rw-r--r-- 1 root root     912 Jan  6 20:33 8a798c20-4e9c-4d10-9c96-616d9ed6ea30.lance
-rw-r--r-- 1 root root 2100796 Jan  6 20:33 8f8c3ce9-bde5-48b8-b6c2-981db57bc034.lance
-rw-r--r-- 1 root root 2100796 Jan  6 20:33 9f7bf18b-4fca-49b9-9931-5e680a709a90.lance
-rw-r--r-- 1 root root 2100796 Jan  6 20:33 b9c1c9f1-1b4d-46cd-baf2-bccb03e9c86d.lance
-rw-r--r-- 1 root root 2100796 Jan  6 20:33 d0304c0c-de39-4fe1-8017-ca8541652d09.lance
-rw-r--r-- 1 root root 2100796 Jan  6 20:33 d197a764-921b-4de0-9527-fc72cf6f9657.lance

after:
# ls -l ./data
total 41028
-rw-r--r-- 1 root root  2100796 Jan  6 20:33 0c602541-d597-4309-9960-f864b9c18d37.lance
-rw-r--r-- 1 root root  2100796 Jan  6 20:33 1004682c-c808-4f22-a9e6-bfb74ca0214c.lance
-rw-r--r-- 1 root root  2100796 Jan  6 20:33 587e1021-9d31-4865-aa36-358ba64e777a.lance
-rw-r--r-- 1 root root  2100796 Jan  6 20:33 628aa704-5647-4138-87ce-4d777a20e20d.lance
-rw-r--r-- 1 root root  2100796 Jan  6 20:33 79bc4fcf-f7c2-4910-b3e9-2a3a78ab34f1.lance
-rw-r--r-- 1 root root  2100796 Jan  6 20:33 8f8c3ce9-bde5-48b8-b6c2-981db57bc034.lance
-rw-r--r-- 1 root root  2100796 Jan  6 20:33 9f7bf18b-4fca-49b9-9931-5e680a709a90.lance
-rw-r--r-- 1 root root 20999113 Jan  6 20:37 a7d0dc0a-b160-40f6-a12b-ac5cda65f463.lance
-rw-r--r-- 1 root root  2100796 Jan  6 20:33 b9c1c9f1-1b4d-46cd-baf2-bccb03e9c86d.lance
-rw-r--r-- 1 root root  2100796 Jan  6 20:33 d0304c0c-de39-4fe1-8017-ca8541652d09.lance
-rw-r--r-- 1 root root  2100796 Jan  6 20:33 d197a764-921b-4de0-9527-fc72cf6f9657.lance
